### PR TITLE
Add a dedicated OpenAI-compatible LLM adapter

### PR DESCRIPTION
**Pull request by jimmyzhuu**
## Summary

This PR adds a dedicated `OpenAI Compatible` LLM adapter for OpenAI-style chat completion endpoints that are not the official OpenAI service.

The implementation is intentionally small in scope:
- adds a new `OpenAI Compatible` LLM adapter backed by LiteLLM's `custom_openai` path
- keeps the existing `OpenAI` adapter unchanged
- adds adapter registration and JSON schema
- adds focused tests for registration, model normalization, schema loading, and usage recording
- documents the new adapter in the README

## Why

Users may already have access to OpenAI-compatible endpoints behind a private gateway or third-party provider, but the current `OpenAI` adapter is specifically shaped around official OpenAI semantics.

Using a separate adapter keeps those semantics explicit and avoids broadening the meaning of the existing `OpenAI` adapter.

Refs #1894
Refs #856
Refs #1443

## Scope

This PR is limited to:
- LLM only
- no embedding changes
- no x2text / OCR changes
- no changes to the existing `OpenAI` adapter behavior

## Notes

`LLM._record_usage` now prefers provider-reported `prompt_tokens` when they are present in the usage payload.

If `prompt_tokens` are missing, `_record_usage` still falls back to LiteLLM token estimation. If that estimation raises, it now logs a warning and records `0` prompt tokens for usage audit instead of bubbling the exception up after a successful LLM call.

This behavior change is in the shared `_record_usage` path, so it applies to every SDK1 LLM adapter that uses it, not just `custom_openai`.

This keeps successful LLM calls from failing at the usage-audit / billing step while preserving the current pricing semantics in this PR.

## Validation

- `UV_SKIP_WHEEL_FILENAME_CHECK=1 uv run pytest tests/test_openai_compatible_adapter.py`
- `UV_SKIP_WHEEL_FILENAME_CHECK=1 uv run ruff check src/unstract/sdk1/adapters/base1.py src/unstract/sdk1/adapters/llm1/__init__.py src/unstract/sdk1/adapters/llm1/openai_compatible.py src/unstract/sdk1/llm.py tests/test_openai_compatible_adapter.py`

### CODE REVIEW

**Comment by jimmyzhuu** [comment]
Addressed the review follow-ups.

- allowed `api_key = null` in the OpenAI-compatible schema
- avoided redundant `token_counter()` calls when provider usage already includes `prompt_tokens`
- tightened the tests around both cases
- added ERNIE / Baidu Qianfan as schema examples while keeping the adapter generic

Validation re-run:
- `UV_SKIP_WHEEL_FILENAME_CHECK=1 uv run pytest tests/test_openai_compatible_adapter.py`
- `UV_SKIP_WHEEL_FILENAME_CHECK=1 uv run ruff check src/unstract/sdk1/llm.py tests/test_openai_compatible_adapter.py`

**Comment by jimmyzhuu** [comment]
Gentle follow-up on this PR in case it slipped through the queue. When someone has bandwidth, I would really appreciate a review. Happy to make any follow-up changes quickly. Thanks!

File: unstract/sdk1/src/unstract/sdk1/llm.py [hunk 1]
```diff
@@ -539,10 +539,21 @@ def _record_usage(
...
+            try:
+                prompt_tokens = token_counter(model=model, messages=messages)
+            except Exception as e:
+                prompt_tokens = 0
+                logger.warning(
+                    "[sdk1][LLM][%s][%s] Failed to estimate prompt tokens: %s",
+                    model,
+                    llm_api,
+                    e,
+                )
```
Comment (ritwik-g) [inline unstract/sdk1/src/unstract/sdk1/llm.py]: @pk-zipstack @johnyrahul is this a safe change?

File: unstract/sdk1/src/unstract/sdk1/llm.py [hunk 2]
```diff
@@ -539,10 +539,21 @@ def _record_usage(
         usage: Mapping[str, int] | None,
         llm_api: str,
     ) -> None:
-        prompt_tokens = token_counter(model=model, messages=messages)
         usage_data: Mapping[str, int] = usage or {}
+        prompt_tokens = usage_data.get("prompt_tokens")
+        if prompt_tokens is None:
+            try:
+                prompt_tokens = token_counter(model=model, messages=messages)
+            except Exception as e:
```
Comment (jaseemjaskp) [inline unstract/sdk1/src/unstract/sdk1/llm.py]: **[High] Silent under-reporting risk — broad `except Exception` + misleading log message.**

This block catches every failure from `token_counter` (unmapped-model `KeyError`, `TypeError`, tokenizer download/IO errors, even programming bugs in litellm) and sets `prompt_tokens = 0`. Audit at line 564 still runs, so billing silently records zero prompt tokens.

The log message (`"Failed to estimate prompt tokens: %s"`) also hides the consequence — a reader scanning logs may assume no usage was recorded, when in fact a zero-prompt-token row was pushed to Audit.

Suggested fix:
```python
except Exception as e:
    prompt_tokens = 0
    logger.exception(  # preserve traceback for unmapped-model debugging
        "[sdk1][LLM][%s][%s] token_counter failed; recording 0 prompt tokens "
        "(cost will be under-reported): %s",
        model, llm_api, e,
    )
```
If possible, narrow to litellm's specific exceptions (`litellm.exceptions.NotFoundError`, `KeyError`) so unrelated bugs surface instead of being swallowed.

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            "is_active": True,
+        }
+
+    @staticmethod
+    def get_name() -> str:
+        return "OpenAI Compatible"
+
+    @staticmethod
+    def get_description() -> str:
+        return "OpenAI-compatible LLM adapter"
```
Comment (jaseemjaskp) [inline unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py]: **[Medium] Tautological description duplicated between `get_metadata()` and `get_description()`.**

Both `get_metadata()["description"]` (line 18) and `get_description()` return the exact string `"OpenAI-compatible LLM adapter"`, which restates the class name and conveys nothing useful to users selecting an adapter in the UI.

Suggested rewrite:
```python
@staticmethod
def get_description() -> str:
    return (
        "Adapter for servers that implement the OpenAI Chat Completions "
        "API (vLLM, LMStudio, self-hosted gateways, third-party providers). "
        "Use OpenAI for the official OpenAI service."
    )
```
and have `get_metadata()["description"]` return the same string (or drop one of the two to avoid drift).

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    @staticmethod
+    def get_provider() -> str:
+        return "custom_openai"
+
+    @staticmethod
+    def get_icon() -> str:
+        return "/icons/adapter-icons/OpenAI.png"
```
Comment (jaseemjaskp) [inline unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py]: **[Low] Icon reused from the `OpenAI` adapter.**

Path `/icons/adapter-icons/OpenAI.png` is also used by `OpenAILLMAdapter`, so the two adapters are visually indistinguishable in the UI — which is exactly the confusion this PR is trying to avoid at the type level.

Not a blocker, but consider adding a dedicated icon (e.g. `OpenAICompatible.png`) so users can tell the adapters apart at a glance.

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      "description": "Provide a unique name for this adapter instance. Example: compatible-gateway-1"
+    },
+    "api_key": {
+      "type": [
+        "string",
+        "null"
+      ],
+      "title": "API Key",
+      "format": "password",
+      "description": "API key for your OpenAI-compatible endpoint. Leave empty if the endpoint does not require one."
```
Comment (jaseemjaskp) [inline unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json:22]: **[Medium] `api_key` misconfiguration is indistinguishable from intentional no-auth.**

`"type": ["string", "null"]` with no `minLength` and no explicit toggle means:
- A user who forgets to paste their key gets a 401 from the provider at request time, not a validation error.
- `null`, `""`, and `"   "` are all accepted and functionally identical.

Consider either:
- Adding an explicit boolean flag (e.g. `requires_auth`, default `true`) so the schema can reject empty `api_key` when auth is expected; or
- Tightening validation in `OpenAICompatibleLLMParameters` to coerce `""` → `None` and treat `None` as "keyless endpoint" explicitly.

As-is the description ("Leave empty if the endpoint does not require one") silently papers over a common misconfiguration.

File: unstract/sdk1/tests/test_openai_compatible_adapter.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            # Stub python-magic so importing LLM does not depend on libmagic
+            # being available in the test environment.
+            "magic": ModuleType("magic")
+        },
+    ):
+        return import_module("unstract.sdk1.llm")
+
+
+def _load_llm_class() -> type:
+    return _load_llm_module().LLM
```
Comment (jaseemjaskp) [inline unstract/sdk1/tests/test_openai_compatible_adapter.py]: **[Low] `_load_llm_class` is dead code.**

Defined but never called — both tests reach in via `_load_llm_module().LLM` directly. Remove it, or use it from both tests for consistency.

**Review by jaseemjaskp — COMMENTED** [review]
## PR Review Toolkit — consolidated findings

Automated review aggregating six specialist agents (code-reviewer, code-simplifier, silent-failure-hunter, type-design-analyzer, pr-test-analyzer, comment-analyzer). No blocking defects; the adapter follows existing sibling-adapter conventions and the scope is appropriately narrow.

**High-signal items worth addressing before merge**
- `llm.py:556` — redundant/confusing fallback: `prompt_tokens` is already resolved at line 543 with an explicit estimation branch, so the `usage_data.get("prompt_tokens", prompt_tokens or 0)` expression double-handles the default and, worse, silently coerces an explicit `None` from `usage_data` to 0 without logging.
- `llm.py:547` — broad `except Exception` combined with a warning that says "failed to estimate" but not "recording 0 tokens" means billing/audit can silently under-report. Narrow the exception and either use `logger.exception` or rewrite the message to name the consequence.
- `base1.py:232` / schema `api_base` — the Pydantic type is plain `str`; URL shape lives only in the JSON schema, so direct construction accepts garbage. Consider `HttpUrl` / a `field_validator`.
- `base1.py:239` — prefix logic is only invoked by the `validate` classmethod. `AzureOpenAILLMParameters` uses `@model_validator(mode="before")` which cannot be bypassed; mirroring that tightens the invariant.
- `custom_openai.json` — `api_base` is listed as required yet ships with a placeholder default URL, which lets users save an unchanged form and hit 404s at request time instead of validation errors. Vendor-specific examples (`ERNIE-4.0-8K (Baidu Qianfan)`, `qianfan.baidubce.com`) are prone to rot and should be generic.
- `test_openai_compatible_adapter.py` — the "tolerates unmapped models" test only asserts `push_usage_data.assert_called_once()`; it does not verify `prompt_tokens=0` was actually pushed, so a regression silently pushing `None` or crashing inside `TokenCounterCompat` would still pass.

**Suggested follow-ups (non-blocking)**
- Add tests for: `api_base` missing (Pydantic ValidationError), `usage=None`, `usage["prompt_tokens"]=None`, and the success branch of `token_counter`.
- Drop the unused `@lru_cache` on `_load_llm_module` and the dead `_load_llm_class`.
- Replace the tautological `get_description()` / metadata description with user-facing copy that distinguishes this adapter from `OpenAILLMAdapter`.

Inline comments below flag each item at its exact line with a concrete fix suggestion.

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json (hunk 5 above)
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json:22]: Added the smaller validation step here: blank or whitespace-only api_key values are now normalized to None, so the keyless-endpoint case stays explicit without expanding the schema.

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py (hunk 4 above)
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py]: Added a dedicated icon and switched the adapter to /icons/adapter-icons/OpenAICompatible.png.

File: unstract/sdk1/src/unstract/sdk1/llm.py (hunk 1 above)
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/llm.py]: Kept this scoped to usage accounting only. It still uses provider-reported prompt tokens when available, only estimates when they are missing, and the fallback paths are covered by tests now.

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py (hunk 3 above)
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/adapters/llm1/openai_compatible.py]: Updated both metadata and get_description() to use user-facing copy that clearly distinguishes this from the official OpenAI adapter.

**Comment by athul-rs** [comment]
Hi @jimmyzhuu - sorry we let this sit so long without a review, that's on us. The change looks reasonable in scope and the validation looks solid. If you're still interested, we'd be happy to have you reopen it and we'll get a maintainer on it this week. Either way, thanks for the careful work and the patience.
cc: @hari-kuriakose @ritwik-g @jaseemjaskp

**Comment by jimmyzhuu** [comment]
Absolutely — I’d be happy to work on this. Thanks for the suggestion!

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    },
+    "model": {
+      "type": "string",
+      "title": "Model",
+      "default": "gpt-4o-mini",
```
Comment (chandrasekharan-zipstack) [inline unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json]:
```suggestion
```
NIT: Consider removing this default since it might eventually get deprecated

**Review by chandrasekharan-zipstack — APPROVED** [review]
LGTM for the most part, @pk-zipstack please help take a look as well

**Review by pk-zipstack — COMMENTED** [review]
LGTM

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json [hunk 8]
```diff
@@ -0,0 +1,61 @@
+{
+  "title": "OpenAI Compatible LLM",
```
Comment (hari-kuriakose) [inline unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json]: @jimmyzhuu This could be just `OpenAI Compatible`.

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json
Comment (hari-kuriakose) [inline unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json:1]: @jimmyzhuu Might be better to rename the file also to `openai_compatible.json`.

**Review by jaseemjaskp — APPROVED** [review]
LGTM. After running the PR review toolkit (code-reviewer, comment-analyzer, pr-test-analyzer, silent-failure-hunter, type-design-analyzer, code-simplifier), all clearly important findings (correctness, billing/audit data integrity, broken contracts) were already raised by prior reviewers (greptile, coderabbitai, jaseemjaskp) and addressed by the author in the latest commits. Remaining items from agent passes are minor or NIT-level (e.g. tightening api_base/model pydantic validators, schema title/filename naming, end-to-end LLM(__init__) coverage), so I'm not posting them as inline comments. Resolved my three previously-posted threads that the current code addresses (DESCRIPTION constant, dedicated icon, blank api_key coercion).

**Comment by athul-rs** [comment]
@jimmyzhuu The one thing I'd note is that the exception-swallowing path now applies to every adapter, not just custom_openai. I think that's the right tradeoff (a successful LLM call shouldn't fail at the billing step), but flagging it for the merge commit.
Could you call out the _record_usage change explicitly in the PR description? 
cc: @hari-kuriakose @pk-zipstack

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json (hunk 7 above)
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json]: @chandrasekharan-zipstack fixed

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json (hunk 8 above)
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json]: @hari-kuriakose fixed.

File: unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/custom_openai.json:1]: @hari-kuriakose fixed. Renamed the schema file to `openai_compatible.json` and updated the adapter schema loading path accordingly.

**Comment by jimmyzhuu** [comment]
> @jimmyzhuu The one thing I'd note is that the exception-swallowing path now applies to every adapter, not just custom_openai. I think that's the right tradeoff (a successful LLM call shouldn't fail at the billing step), but flagging it for the merge commit.我注意到的一点是，现在异常捕获路径适用于所有适配器，而不仅仅是 custom_openai。我认为这是正确的权衡（一个成功的 LLM 调用不应该在计费步骤失败），但为合并提交标记一下。 Could you call out the _record_usage change explicitly in the PR description?你能明确在 PR 描述中提到 _record_usage 的变更吗？ cc: @hari-kuriakose @pk-zipstack

@athul-rs Updated the PR description.

File: unstract/sdk1/src/unstract/sdk1/llm.py [hunk 9]
```diff
@@ -539,10 +539,22 @@ def _record_usage(
...
+                    model,
+                    llm_api,
+                    e,
+                )
         all_tokens = TokenCounterCompat(
-            prompt_tokens=usage_data.get("prompt_tokens", 0),
+            prompt_tokens=prompt_tokens or 0,
             completion_tokens=usage_data.get("completion_tokens", 0),
             total_tokens=usage_data.get("total_tokens", 0),
         )
```
[the remark this answers is not in the document]
Comment (ritwik-g) [reply unstract/sdk1/src/unstract/sdk1/llm.py]: @jimmyzhuu could you take a look at this comment?

File: unstract/sdk1/src/unstract/sdk1/llm.py (hunk 1 above)
Comment (praveen-formido) [reply unstract/sdk1/src/unstract/sdk1/llm.py]: LGTM, safe to merge. Reported prompt_tokens still wins for all major providers (OpenAI, Anthropic, Bedrock, Gemini, etc.), only the fallback path changes. Also fixes a latent bug where prompt_tokens: None was passed straight into TokenCounterCompat, and avoids crashing _record_usage on unmapped models. Tests cover all three paths.

File: unstract/sdk1/src/unstract/sdk1/llm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jimmyzhuu) [reply unstract/sdk1/src/unstract/sdk1/llm.py]: > @jimmyzhuu could you take a look at this comment?你能看看这条评论吗？

Kept this fallback intentional. The audit-fidelity risk is real, but for this PR we chose not to let post-success token estimation failures fail an otherwise successful LLM call. Provider-reported `prompt_tokens` still win when present; only the fallback path changes. I think provenance fields like `prompt_tokens_source` / `estimation_failed` belong in a follow-up, since they need a wider end-to-end usage contract change beyond sdk1.

[the remark this answers is not in the document]
Comment (ritwik-g) [reply unstract/sdk1/src/unstract/sdk1/llm.py]: @hari-kuriakose @coderabbitai Can we move forward with this?

[the remark this answers is not in the document]
Comment (hari-kuriakose) [reply unstract/sdk1/src/unstract/sdk1/llm.py]: @coderabbitai Yes, please open a follow-up GitHub issue to track adding a provenance field to the audit payload so downstream consumers can distinguish missing-data from genuinely-zero usage.

File: unstract/sdk1/src/unstract/sdk1/adapters/base1.py [hunk 10]
```diff
@@ -76,6 +76,8 @@ def register_adapters(adapters: dict[str, dict[str, "Any"]], adapter_type: str)
 class BaseAdapter(ABC):
     """Adapter base class for compatibility with all SDK v1 providers."""
 
+    SCHEMA_PATH: ClassVar[str | None] = None
```
Comment (ritwik-g) [inline unstract/sdk1/src/unstract/sdk1/adapters/base1.py]: @jimmyzhuu @pk-zipstack may be a silly question. Why we had to refactor  SCHEMA_PATH as a class variable? I was just wondering if using like this could have some other side effect since we use BaseAdapter as baseclass for other adapter classes? 

I mean wouldn't renaming the `openai_comptatible.json` to `custom_openai.json` have avoided these changes all together. 

@jimmyzhuu If the above works I would suggest doing that and reverting the introduced logic for cls.SCHEMA_PATH. While nothing inherently wrong in doing what you did we want to keep the changes to based class as minimal and none when possible.

Comment (ritwik-g) [reply unstract/sdk1/src/unstract/sdk1/adapters/base1.py]: cc: @greptileai

Comment (ritwik-g) [reply unstract/sdk1/src/unstract/sdk1/adapters/base1.py]: @coderabbitai can you raise this fix?

**Comment by athul-rs** [comment]
Hi @jimmyzhuu  — thanks for the PR! I pushed a small follow-up commit (27215c2d) addressing @ritwik-g's 
  ▎ review on base1.py.
  ▎
  ▎ The change keeps BaseAdapter untouched by aligning with the convention every other adapter follows: name 
  ▎ the schema file after get_provider(). Since OpenAICompatibleLLMAdapter.get_provider() returns 
  ▎ "custom_openai", renaming the JSON from openai_compatible.json → custom_openai.json lets the default 
  ▎ get_json_schema() path resolution find it, so the introduced SCHEMA_PATH class variable and the if 
  ▎ schema_path is None branch are no longer needed.
  ▎
  ▎ Specifically:
  ▎ - Renamed unstract/sdk1/src/unstract/sdk1/adapters/llm1/static/openai_compatible.json → custom_openai.json
  ▎ - Dropped SCHEMA_PATH and the ClassVar import from BaseAdapter; simplified get_json_schema() back to the 
  ▎ default path build
  ▎ - Dropped the SCHEMA_PATH override and unused os / ClassVar imports from OpenAICompatibleLLMAdapter
  ▎ - Updated test_openai_compatible_schema_is_loadable to read via get_json_schema() instead of touching 
  ▎ SCHEMA_PATH directly

File: README.md [hunk 11]
```diff
@@ -176,9 +176,10 @@ Also see [architecture](docs/ARCHITECTURE.md).
 | Provider | Status | Provider | Status |
 |----------|--------|----------|--------|
 | OpenAI | ✅ | Azure OpenAI | ✅ |
-| Anthropic Claude | ✅ | Google Gemini | ✅ |
-| AWS Bedrock | ✅ | Mistral AI | ✅ |
-| Ollama (local) | ✅ | Anyscale | ✅ |
+| OpenAI Compatible | ✅ | Anthropic Claude | ✅ |
```
Comment (ritwik-g) [inline README.md:179]: @hari-kuriakose is this provider correct?

**Review by ritwik-g — APPROVED** [review]

**Review by pk-zipstack — APPROVED** [review]
